### DISCUSSION TITLE
Specifying user password

### DISCUSSION THREAD

**Discussion by dshaw619** [gitbls/sdm#41] [Ideas]
I've noticed that if the user is specified during customization, but not a password switch, I am prompted for a password, which is great because the password doesn't have to be stored in the script file I typically use to execute sdm.  This doesn't work during --burn though. 

I generally use different passwords on each of my Pi's so ideally wouldn't need to specify one during customization, but only during the burn.  I can use a default password for the customization and then override it during the burn, but have to include it on the command line (I found out the hard way what happens if I just include the --password-user switch with no password parameter (:-) ).  

I have scripts from which I like to execute sdm so I'm sure I'm using all the necessary parameters, but it would be bad practice to store passwords in such a script (and a nuisance to have to edit it for each burn of a different system).  I've been copying the burn command from my script file, pasting it into the terminal, and then editing in the password copied from my password manager, but it would be much more convenient if the --burn process prompted for an unspecified password like --customize.

**Comment by gitbls** [comment]
Interesting point. While it's possible to write a temporary script that has the burn command (plus password) in it, which you'd then delete after the burn, that still leaves the password exposed in a (presumably appropriately protected) script file during the long-running burn.

I'll have a look into this as well. Haven't thought much about it, but may need to add a switch to request password prompting. 

Would that work for you?

**Reply by dshaw619** [reply]
That (adding a switch to request password prompting) would work well.  Or even prompting for it if --user was specified in the burn command (if you do add that - low priority) and the password was not given in the command line.

**Comment by gitbls** [comment]
Current thinking is to add `--user` to the burn command. `--password-user secretpwd` will be accepted as well, and if not specified along with `--user`, it will prompt for the password. There will be no way to add a user with no password using the `--user` switch at burn time without being prompted. Haven't decided yet whether entering a null password when prompted is OK.

If a user is created during customization with `--user`, nothing described above would modify that user.

**Comment by gitbls** [comment]
V7.2 implements this. You must enter a password, a null password is not allowed.

**Reply by dshaw619** [reply]
I've tested the new v7.2 and it works successfully.   However, rereading https://github.com/gitbls/sdm/wiki/Passwords, I see that the password is stored in plaintext in /etc/sdm/cparams even if not entered on the command line, but prompted.  I double-checked and see that that is the case.  Rather than counting on remembering to delete this file on the burned image, I'll probably adopt a default initial password and then change it outside of sdm after the SD card is first booted.

**Comment by gitbls** [comment]
I've been thinking about this as well, although not an burning (🤣) issue for me personally, I've not been totally happy with leaving the passwords there.

It looks pretty simple to redact the passwords from `history` and `cparams`, driven by a command-line switch, 

I'll get this in soon...likely next week.

**Comment by gitbls** [comment]
Simple (context-free) reaction is in V7.3, posted this afternoon.

**Reply by dshaw619** [reply]
--redact works nicely.  Thank you.

**Comment by gitbls** [comment]
For completeness, note that /etc/sdm is protected 700, so non-root users can't casually access it without `sudo`. Of course, if an account can `sudo`, it has access to the whole system.
